async await codelab

legalcodes opened the pull request:
Staged:
https://jt-ww-1.firebaseapp.com/codelabs/async-await

Remaining tasks: (last updated 7/19)

- [x] Incorporate changes from Round 2 UX studies
- [x] observe test user completing the codelab
- [x] Incorporate changes needed after observing tester
- [x] Submit for final review

RedBrogdon commented:
Feel free to ping me if I can be of use when you're building the gists for this.

kwalrath commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+---
+
```
We'll need the title and description here.

kwalrath commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+
+
+Welcome to the asynchronous Dart codelab! In this codelab you'll practice using Dart to run asynchronous code via the `async` and `await` keywords. 
```
Let's have 80-character lines (or shorter).

I like breaking on phrases (roughly), for shorter diffs in future changes. If you do that, take care so that reviewers know there's a continuation, if necessary. E.g., "... and _clause_" would be broken into two lines like "... and" & "_clause_" instead of "..." & "and _clause_".

kwalrath commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+### **Lifecycle of a Future Object**
+
+### **Uncompleted**
```
The headings shouldn't have asterisks.

kwalrath commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+
+
+## Asynchronous Dart Code: Future, async, and await
```
We use sentence case in our titles and headings. https://developers.google.com/style/capitalization#capitalization-in-titles-and-headings

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+The try-catch code behaves in the same way with asynchronous code as it does with synchronous code: if the code within the `try` block throws an exception, the code inside the `catch` clause executes. 
+
+
+## Practice Handling Errors with async and awaitImplement code to accomplish the following:
```
A header shouldn't have punctuation at the end. So:

## Practice handling errors 

Use `async` and `awaitImplement` to accomplish the following:

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+
+
+* Make `changeUsername` call the provided asynchronous function `getNewUsername`, and return its result
```
* Update `changeUsername` to call the provided asynchronous function `getNewUsername`, and return
  its result.  (Add a period)
*

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+
+* Make `changeUsername` call the provided asynchronous function `getNewUsername`, and return its result
+* `changeUsername` must catch and return any errors thrown by getNewUsername, prefixing the returned error with the following string: `Error:` 
```
=> The `changeUsername` method must catch and return any errors thrown by `getNewUsername`.
      Prefix the returned error with the string `Error:`.

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+* Example return value from `changeUsername` when `getNewUsername` throws an error: `Error: Username must contain only alphanumeric values`
+
+
+## Putting It All Together
```
=> Putting it all together  (sentence case)

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+
+
+Now it's time to sum up what you've learned in one final exercise. In the following snippet, write the following:
```
=> In the following example, do the following tasks:

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+Find Out More
+
+*  [Futures](https://www.dartlang.org/tutorials/language/futures)
```
* The [Futures]() tutorial.

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+Find Out More
+
+*  [Futures](https://www.dartlang.org/tutorials/language/futures)
+*  [The Event Loop in Dart](https://webdev.dartlang.org/articles/performance/event-loop)
```
The Event Loop in Dart article is out of date. Ask Kathy if you should link to it.

kwalrath replied:
Avoid linking to it unless you really think it's important.
Also... check the URLs. webdev.dartlang.org doesn't exist any more, except to redirect.

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+*  [Futures](https://www.dartlang.org/tutorials/language/futures)
+*  [The Event Loop in Dart](https://webdev.dartlang.org/articles/performance/event-loop)
+*  [Isolates](https://api.dartlang.org/stable/2.2.0/dart-isolate/dart-isolate-library.html)
```
Use a release agnostic link for APIs. We actually don't have a lot of useful info about isolates, due to historical reasons. Also, use the {{site.dart.api}} macro.

sfshaza2 commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+*  [Futures](https://www.dartlang.org/tutorials/language/futures)
+*  [The Event Loop in Dart](https://webdev.dartlang.org/articles/performance/event-loop)
+*  [Isolates](https://api.dartlang.org/stable/2.2.0/dart-isolate/dart-isolate-library.html)
+* Read about  [Dart's type system](https://dart.dev/guides/language/sound-dart) for more examples of type signatures with futures. You should  [prefer signatures in function type annotations](https://dart.dev/guides/language/effective-dart/design#prefer-signatures-in-function-type-annotations) wherever possible!
```
Use relative links, not absolute links, when possible. In this case:

[Dart's type system](/guides/lanuage/sound-dart) for ....

I know that you link to Effective Dart for "preferring signatures", but can you summarize, in one sentence, why? I clicked over to the link and it's not really clear why. (And if you can't, maybe this advice should be deleted here.)

kwalrath replied:
I'm not sure that "relative" is the right term here, but yep, never use "dart.dev" for anything in dart.dev. If you think the page might move, you could use a macro instead (`{{site.dartlang}}` or  something like that.)

kwalrath commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+`// Your order is: Instance of '_Future<String>'`
+
+* after 1 second has passed: the future is uncomplete - we cannot access the desired result yet
```
Actually, I believe we don't say "incomplete", wrt Futures. Instead, we might describe it as an "uncompleted future". https://dart.dev/tutorials/language/futures sets the standard.

sfshaza2 replied:
I wondered about that. Thanks for jumping in. I can live with "uncompleted future".  :D

legalcodes commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+
+
+In the previous section you learned that, in Dart, future objects represent the results of asynchronous operations -- if a function does asynchronous work, it returns a future object. But how do we access the results of future objects when using asynchronous functions? 
```
`But how to access the results of future objects when using asynchronous functions?`
Compare with:
`But how to walk when using legs?`
Not really a full sentence (missing subject)

legalcodes commented on src/_guides/language/async-await.md:
```diff
@@ -0,0 +1,344 @@
+
+    Future<String> createOrderMessage () async {
+
+ *Note: You can *  [use Future<void>](https://dart.dev/guides/language/effective-dart/design#do-use-futurevoid-as-the-return-type-of-asynchronous-members-that-do-not-produce-values) * as the return type for asynchronous functions that don't return usable values. * 
```
I did some digging and ended up with `<div class="alert alert-info" markdown="1">` but feel free to let me know if we have different pattern for this.

sfshaza2 replied:
No, we use a liquid template:

```
{{site.alert.info}}    (I think we have a .info..., otherwise site.alert.note or 
                                 site.alert.secondary, depending)
  Text here
{{site.alert.end}}
```

legalcodes replied:
We've configured these templates in flutter/website, but not in dart-lang/site-www. 

Notice that  after running `jekyll build`, the following :
```
{{ site.alert.info }}
INFO
{{ site.alert.end }}

{{ site.alert.note }}
NOTE
{{ site.alert.end }}

{{ site.alert.secondary }}
SECONDARY
{{ site.alert.end }}
```

renders to:

![Screen Shot 2019-06-05 at 4 51 30 PM](https://user-images.githubusercontent.com/5550175/58997640-5a465d00-87b2-11e9-8dcb-e9f4950d9e9a.png)

legalcodes replied:
I'll update the PR to add these templates to the site's config.yml, then sub out the divs for the templates.

legalcodes replied:
Quick update about these notes:

IMHO these styles don't match the use case I have in mind. I was going for something subtle that didn't draw too much of the the viewer's attention. These notes are really bulky, and after just 2 of them they start swallowing screen real estate and distract the viewer from the main flow of the article/tutorial/codelab:

### Before: 

![Screen Shot 2019-06-07 at 3 26 34 PM](https://user-images.githubusercontent.com/5550175/59136608-ba670b80-8938-11e9-9ce3-2ed06dc9b3be.png)

### After:
![Screen Shot 2019-06-07 at 3 19 29 PM](https://user-images.githubusercontent.com/5550175/59136434-01083600-8938-11e9-8fde-39b290c2a314.png)

Here are some options I have in mind (please chime in if there are others ):
1. We update these styles
2. We add new "templates for "softer" styles if we want to keep existing info alert "loud" styles

legalcodes commented on src/_guides/language/async-await.md:
````diff
@@ -0,0 +1,344 @@
+```
+
+*  *Note: *  [All functions return a value](https://dart.dev/guides/language/language-tour#return-values) * even if you don't explicitly use the *  *`return`*  * keyword. * 
+*  *Note: Dart can *  [infer the Future<String> type](https://dart.dev/guides/language/sound-dart#type-inference) * for you.
````
I worry that `Dart infers the Future type for you` might mislead the reader into thinking that Dart *always* infers the Future type for you. I've linked to the rules regarding type inference so that we don't duplicate efforts across docs. 

For now I've left the `can` in since I think this is somewhat more true to the language. LMK if you'd like to discuss more.

kwalrath commented:
Cool stuff! But of course, I have feedback.

Some general feedback: I was surprised that there was only one embedded DartPad. I expect a lot more (inter)activity in a codelab. This feels more like a tutorial. (Maybe we can't make a codelab out of this; that's fine, a tutorial is good too.)

Whether it's a tutorial or codelab, even if we don't have exercises, we should prefer letting people execute code, as opposed to just telling them what happens when the code executes.

Some more specific feedback:
* We need to be careful about terminology. For example:
  * We shouldn't say that futures return values. Instead, they complete with values (or errors).
  * "`Future` objects" == "futures". We shouldn't use the term "future object".
* We need to be careful about how we use "async" and "asynchrony". I've been using _async function_ to mean a function that has the `async` keyword before its body. As opposed to an _asynchronous function_, which might or might not have the `async` keyword. (It's perhaps too fine a point... but we do need to differentiate somehow, since the behavior of async-await is very different from the behavior of functions that use the Future API instead.) And then there's _asynchrony_ (not _async_) or _asynchronous code/programming/whatever_, which I use to mean the subject in general.
* When I printed the codelab, the font was very small. I think this is due to the embedded DartPad. We should do something to make DartPads not be so wide in the printed version. I think it's a problem we need to solve for mobile, too.
* I found the instructions (and in particular the parts) to be confusing.
* Let's avoid sentence fragments. 
* There are two titles on the page, and then a section head (**Introduction**) right beneath the title. Delete the section head and fix the titles.
* We should be clear about the prerequisites (especially if this is a codelab; less of an issue if this is a tutorial). E.g., do they need experience writing/reading Dart code?
* The timeline under the "What is a future" section is hard to read. Use a table or a figure?
* I was surprised by the "Quick Review". That seemed like something you might have at the end of the whole article. But if we have it here, we should have it at the end of every major section.
* Why isn't the code under "Execution flow with async and await" in a DartPad? We could save space _and_ make it interactive!

I didn't want to make line-by-line comments, since it sounds like the content still isn't set. (Is that right?)

legalcodes commented:
` I was surprised that there was only one embedded DartPad.`
There will be 3 total, the Errors section and the Putting it All Together section will both have embedded DartPads.

legalcodes commented:
`we should prefer letting people execute code, as opposed to just telling them what happens when the code executes.`

This is a really good point -- there may be some easy opportunities to replace some of the examples with an embedded DartPad. Will update you on this.

legalcodes commented:
`We should be clear about the prerequisites (especially if this is a codelab; less of an issue if this is a tutorial). E.g., do they need experience writing/reading Dart code?`

The current draft lists only one pre-req (some experience writing asynchronous code) because if you've done that, you really can get through the codelab. Let me know if you think otherwise.

legalcodes commented:
`Why isn't the code under "Execution flow with async and await" in a DartPad? We could save space and make it interactive!`

Similar to above, I think this is a great idea. Will incorporate this.

legalcodes commented:
`I found the instructions (and in particular the parts) to be confusing.`

OK, good to know. The parts are there so that errors from test code output can reference them and give the user a clearer sense of which parts are failing.

galeyang commented:
> `I was surprised that there was only one embedded DartPad.`
> There will be 3 total, the Errors section and the Putting it All Together section will both have embedded DartPads.

I actually included at least 5 embedded DartPads in the study. Considering that I didn't include all sections in the study, I would expect to have probably 8 embeds in your full version. Feel free to ping me offline as needed.

> `We should be clear about the prerequisites (especially if this is a codelab; less of an issue if this is a tutorial). E.g., do they need experience writing/reading Dart code?`
> 
> The current draft lists only one pre-req (some experience writing asynchronous code) because if you've done that, you really can get through the codelab. Let me know if you think otherwise.

Some pre-reqs based on my learning from a recent UX study
1. Have hand-ons programming experience in another language
2. Know little Dart basic syntax
3. Some experience writing asynchronous code in another language

For the point 2, in UX study, I invited developers with "zero Dart experience" but know other programming languages. Before jumping into this codelab, I provided  a cheat sheet that introduces basic syntax used in this codelab. I can share you the cheat sheet I created. Maybe we can offer similar things to new users so that it's easier to get started (without getting lost in a tour of language doc).

> (Re: K) When I printed the codelab, the font was very small. I think this is due to the embedded DartPad. We should do something to make DartPads not be so wide in the printed version. I think it's a problem we need to solve for mobile, too. 

The font size on mobile looks fine to me but the experience can be better. For printed version, I guess current Dartpad2 is not optimized for printing. It looks very confusing now.

> (Re: K) I was surprised by the "Quick Review". That seemed like something you might have at the end of the whole article. But if we have it here, we should have it at the end of every major section.
+1 to this.

kwalrath commented:
> `We should be clear about the prerequisites (especially if this is a codelab; less of an issue if this is a tutorial). E.g., do they need experience writing/reading Dart code?`
> 
> The current draft lists only one pre-req (some experience writing asynchronous code) because if you've done that, you really can get through the codelab. Let me know if you think otherwise.

If they don't need Dart experience, we should say that.

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -312,32 +313,49 @@ main() async {
+1. `getLoginAmount` returns an `int` representing the number of times that the user has logged in.
+1. Example return value from `reportLogins: 'Total number of logins: 57'`
 
+<iframe frameborder="no" height="525" src="https://dartpad2-ux.firebaseapp.com/experimental/embed-new?id=f751b692502c4ee43d932f745860b056" width="100%"></iframe>
```
Use production links for Dartpad2 instead of the UX experimental site. 
https://dartpad.dartlang.org/experimental/embed-new.html?id=f751b692502c4ee43d932f745860b056

kwalrath commented:
> For the point 2, in UX study, I invited developers with "zero Dart experience" but know other programming languages. Before jumping into this codelab, I provided a cheat sheet that introduces basic syntax used in this codelab. I can share you the cheat sheet I created. Maybe we can offer similar things to new users so that it's easier to get started (without getting lost in a tour of language doc).

We have a cheatsheet on the site, but I haven't linked to it, because I didn't feel like it was ready yet (details at https://github.com/dart-lang/site-www/pull/1625#issuecomment-497373967) . Maybe we can merge it and your cheatsheet?

Here's the semi-published cheatsheet: https://dart.dev/guides/language/cheatsheet

legalcodes commented:
Sounds good, I'll use the pre-reqs that Gale drew from the studies (including needing at least a little bit of dart experience). I'll also take a closer look at the various cheatsheet options we've proposed.

legalcodes commented:
Staged:
https://jt-ww-1.firebaseapp.com/codelabs/async-await

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+
+## Introduction
+ 
+Welcome to the asynchronous Dart codelab! In this codelab you practice using
```
In this codelab**,** you practice using Dart to run asynchronous code via the `async` and `await` keywords.

-three or more rule for prep phrases. i read this and expected **you'll** not **you*, but **you* makes sense; it's just hard for me to process this sentence without needing a short pause.

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+Welcome to the asynchronous Dart codelab! In this codelab you practice using
+Dart to run asynchronous code via the `async` and `await` keywords.
+
+In addition to explanations, this codelab includes embedded editors that show 
```
In addition to explanations **(explanations of what?)**, this codelab includes embedded editors that show 
you working code examples and coding exercises.

-just a question i found myself asking.

legalcodes replied:
Nice, will make this more clear. Thanks for pointing it out.

legalcodes replied:
Yep, this sentence could be more clear. Thanks!

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+
+Before you begin this codelab, you should:
+
+* Have programming experience in another language
```
* Have programming experience in another language **(what qualifies experience?)**
* Know basic **Dart syntax**
* Have some experience writing asynchronous code in another language

-as a noob, i'm curious about your target audience's experience level, or what your expectations are for them. For clarity (and also concision), **Dart** sounds better as an adjective than an object of a prep.

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+
+When you finish this codelab, you'll know the following: 
+
+* How to use the `async` keyword
```
- I like this list; if I were more tech savvy, I'd be pretty psyched to work through this codelab; it seems like I'll gain more than just additional tech skills.

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+* Know basic syntax in Dart 
+* Have some experience writing asynchronous code in another language
+
+When you finish this codelab, you'll know the following: 
```
When you finish this codelab, you'll **(you will)** know the following:

-not sure what your stance is on contractions, but I noticed this is the only one for **you will** in your documentation; you spell out **you will** the other two times you combine the subject *you* and modal verb *will*; maybe for consistency spell out **you'll** in this sentence.

legalcodes replied:
I'm following [the style guide](https://developers.google.com/style/contractions), so feel free to point out anywhere I've missed it. 

If anything I'll try to change from future to [present tense](https://developers.google.com/style/tense) in this line (L23)

legalcodes replied:
`This codelab covers the following material:`

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+* Writing to a database
+* Reading data from a file
+
+To implement these features, you will need to learn the Dart syntax for
```
To implement these features, you will need to learn the Dart syntax for
futures, async **(`async`)**, and await **(`await`)**.

-do you not need backticks around these words?

legalcodes replied:
I was wondering about this myself -- it felt awkward monospacing async/await but not futures:

...learn the Dart syntax for futures, `async`, and `await`

But then again, it's wrong to monospace futures:

...learn the Dart syntax for `futures`, `async`, and `await`

Since the word 'futures' isn't a language keyword or a classname. `Future` is of course a classname: 

...learn the Dart syntax for the `Future` class, and `async` and `await` keywords

But that sounds more verbose than it needs to be for a single sentence introduction.

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+
+In the above code:
+
+* `getUserOrder` is an asynchronous function that waits four seconds before
```
* `getUserOrder` is an asynchronous function that waits four seconds before
returning a string**.** **In this example, the string describes** the user's order: a large latte. (For a further
explanation you can [find out more about
Future.delayed]({{site.dart_api}}/Future.delayed.html))**.**
-you have a long but grammatically correct sentence; good usage of relatives clauses and gerunds. I'd break this sentence up, though, for readability. 
* Since `getUserOrder` waits four seconds before returning the user's order,
`createOrderMessage` never **~~actually~~** prints the user's order. Instead the console
prints the value immediately returned by `createOrderMessage`**,** which
turns out to be `Instance of '_Future<String>'`.
-this is another long but good sentence; however, i'd cut out **actually** because it's an empty adverb here. I'd also add a comma before the relative pronoun **which**.

legalcodes replied:
- Agreed: I should break up this long sentence
- I've cut out actually, nice catch.

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+prints the value immediately returned by `createOrderMessage` which
+turns out to be `Instance of '_Future<String>'`.
+
+What is a future, and how do you write code to handle the asynchronous function
```
What is a future, and how do you write code to handle the asynchronous function
 `getUserOrder`?

**wait, I'm lost. you ask what is a future above, and then below (in quick review) you go over how Futures write declarative asynchronous code. is a future different from Futures? are they terms uses synonymously? did you go over Futures before quick review?**

 {{ site.alert.secondary }} 
 Quick review: 
 * Asynchronous operations let a program do work while waiting for an operation to finish.
 * In Dart, you can use Futures, `async` and `await` to write declarative asynchronous code.
 {{ site.alert.end }}

legalcodes replied:
I'll remove mention of futures from the quick review. They're covered in the next section.

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+ {{ site.alert.end }}
+
+
+## Async in Dart: What is a future?
```
## Async in Dart: What is a future?

In Dart, a
[Future]({{site.dart_api}}/Future-class.html)
represents the result of an asynchronous operation. A future object has
two states: uncompleted or completed.
-should **future** in the header and **Future** in the text both be capitalized?

legalcodes replied:
See the [style guide on sentence casing](https://developers.google.com/style/capitalization)

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+represents the result of an asynchronous operation. A future object has
+two states: uncompleted or completed.
+
+{{ site.alert.note }}
```
'Uncompleted' is a ~~special term in~~ **Dart term** referring to the state of a future 
 before it has produced a result.
-i think putting information in a prep phrase sometimes can slow readers down; like one of my earlier comments, I think using Dart as an adjective is more clear because, like **syntax**, **term** is the noun, and **Dart** is used to describe it.

legalcodes replied:
Yeah, there may be better ways to do this. What I'm going for here is this:

I want to introduce readers to the fact that `uncompleted` is a legitimate descriptor in Dart, and not a typo.

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+of asynchronous functions. 
+
+{{ site.alert.secondary }}
+Quick Review:
```
-i  like your quick review; i think I understand what your writing about at a high level. i'm curious, though, about why a Future is called a Future?

legalcodes replied:
The pattern has been around for a while: https://en.wikipedia.org/wiki/Futures_and_promises

While I don't think that reviewing the history would be useful in the codelab, perhaps an additional sentence or two explaining how the name relates to the concept might help?

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+
+## Working with futures in Dart: async and await
+
+The `async` and `await` keywords provide a declarative way to: Define 
```
The `async` and `await` keywords provide a declarative way to~~:~~ **d**efine 
asynchronous functions and access their results---*after they have completed.*

legalcodes replied:
The colon is a typo leftover from when this section used to be in list format. Thanks for the catch!

datafoya commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,346 @@
+ as the return type for asynchronous functions that don't return usable values.
+{{ site.alert.end }}
+
+Now that you've declared an async function, it returns a future object. Use the
```
-reading this sentence makes me think I've been following a how-to manual.

legalcodes replied:
That's probably a good thing -- this is the end of a section describing how to use the async and await keywords. Let me know if you have additional thoughts on this.

galeyang commented:
More notes on the staging version

1. The example above the "after 1s ...& after 4s" part is a bit complicated to understand before I learn rest of the content.  In the previous iteration, you show users correct examples in the next section "Working with futures in Dart: async and await" with the before/after comparison. For me, that's easier to follow.
2.  I like that you can use dartpad in "Working with futures in Dart: async and await". However, the downside is that you can't highlight the difference with the before/after comparison. In the previous study, participants like having before vs. after code examples and highlighting the difference.
3. Execution flow with async and await. This part will need the new UI that show output on the right, otherwise it's hard to refer to output and code. We'd like to ask Andrew about the deployment plan. Alternative for now: add output as comments
4.  (cont. point3 ) Could you add a comment in both code examples to help users quickly identify the difference in both code "the print statement “Awaiting user order” moves to the line after the first await keyword in createOrderMessage"?
![image](https://user-images.githubusercontent.com/5102864/59309089-82710880-8c57-11e9-90a6-705e7c0fb7bf.png)
![image](https://user-images.githubusercontent.com/5102864/59309191-cf54df00-8c57-11e9-83d1-0b0d0ac3f8a6.png)

5. (cont. point3 ) At first glance, I'm not clear if this statement “Notice that the timing of the output shifts if the print statement “Awaiting user order” moves to the line after the first await keyword in createOrderMessage:" is referring to the code above or underneath.

![image](https://user-images.githubusercontent.com/5102864/59308812-d2030480-8c56-11e9-96ad-d14e8c6f29c9.png)

6. Have extra `}` and indentation in this part
![image](https://user-images.githubusercontent.com/5102864/59309584-c9abc900-8c58-11e9-9f03-01c020d9d415.png)

7. "Putting it all together" formatting is difficult to read

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,453 @@
+{{ site.alert.end }}
+
+Now that you have declared an async function:
+  1. Your function returns a future that completes with the type that you specify
```
Add a period.

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,453 @@
+* Read about [Dart's type system](/guides/language/sound-dart) for more examples
+of type signatures with futures. You should
+[prefer signatures in function type annotations](/guides/language/effective-dart/design#prefer-signatures-in-function-type-annotations)
+wherever possible!
```
This list is not very parallel. Also, I'd reorder a tad. Perhaps:

* [Asynchronous programming: futures & async-await]() tutorial
* [Dart's type system][] (Includes more examples of type signatures with futures.)
* [Preferring signatures in function type annotations][], from [Effective Dart][]
* [Dart codelabs][]
* [DartPad][]

sfshaza2 commented on src/codelabs/index.md:
```diff
@@ -20,6 +20,10 @@ Use the latest, experimental version of DartPad to learn,
 remind yourself about, or test your knowledge of
 some of the most commonly used, yet unique features of the Dart language.
 
+### [Futures Async & Await](/codelabs/async-await)
```
You're self linking?

legalcodes replied:
`/codelabs/index.md` is the landing page for the "Codelabs" submenu, and where we are listing the codelabs on dart.dev. Since the futures codelab is a codelab it belongs on this landing page, right? Let me know if I'm missing something

sfshaza2 replied:
D’OH. I missed the filename you were modifying.  :D

legalcodes commented:
@sfShaza2 The staged link is my first comment on this PR (which by now is buried) so I'll add the link in this comment as well as in the PR description:

Staged:
https://jt-ww-1.firebaseapp.com/codelabs/async-await

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,453 @@
+**future**   
+"future" refers to an instance of the Dart Future class.
+
+__`async`__  
```
Although they are code fonted, in this case they're not italicized; they're bolded.

Glad you brought this up, I was ambivalent about it -- this is a case where we would normally code font them, but (as you point out) it looks odd doing so since the other bolded items don't need code font. Since you're OK with leaving them without the code font, I'm happy to do so. Thanks!

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,453 @@
+
+When you invoke a function that returns a future, that function immediately 
+returns a future in an uncompleted state. While in its uncompleted state,
+a future is waiting for asynchronous operations to finish or to 
```
Thanks for bringing this up, I'm also ambivalent about this -- depending on how we want to use the term "operations" of course, the more interesting use cases for async/await **do** involve multiple asynchronous operations. 

e.g.:
```dart
Future<String> getUserAuthMessage(username) async {
  String user = await getUser(username)
  String auth = await authenticate(user);
  String message = await getUserMessage(auth);
  return "$user $auth $message";
} 

String userAuthMessage = getUserAuthMessage();

```

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,452 @@
+{% endprettify %}
+
+
+The following example provides a side-by-side comparison of a synchronous function
```
Wow, this confused me for a bit. By using “example” singular and “side-by-side comparison” I was truly expecting to see both shown in the following DartPad. I kept scanning up and down to figure it out. So, it’s the following “exampleS” and it’s not side-by-side.  :)

I’d change the text to something like: “The following examples provide a comparison of a…” The first example shows synchronous code. The second example modifies the first to demonstrate asynchronous code.

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,452 @@
+The following example provides a side-by-side comparison of a synchronous function
+and an asynchronous function. Since you can only use the `await` keyword within 
+an `async` function body, the example converts the parent function from 
+synchronous to asynchronous thus enabling use of the `await` keyword: 
```
And, since I was already confused by the first sentence, this second sentence made no sense. I would move this thought to be just in front of the 2nd DartPad and say something like:

The following example shows the same code as before, but implemented asynchronously. Since you can only use the `await` keyword within an `async` function body, the `createOrderMessage()` function is now tagged as `async`.

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,452 @@
+
+## Execution flow with async and await
+
+As of Dart 2.0, functions marked as `async` run synchronously until the first
```
We use “Dart 2”, not “Dart 2.0”.  https://docs.google.com/document/d/1vClxUPrWU1fsRq5Ws_KSz6cZHJ2QxkCCG436HJNNGh4/edit#heading=h.3ka2czxdh5qy

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,452 @@
+  width="100%">
+</iframe>
+
+Notice that the timing of the output shifts if the print statement
```
I would suggest that they move the line themselves and delete the following DartPad instance.

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,452 @@
+
+## Putting it all together
+
+It's time to sum up what you've learned in one final exercise. Write the following:
```
sum up => practice

sfshaza2 commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,452 @@
+It's time to sum up what you've learned in one final exercise. Write the following:
+
+### Part 1: `addHello` 
+* Write a function `addHello.` that takes a single string
```
Delete the period after “addHello”.

legalcodes commented:
Thanks for these updates!

`This should be battle tested before we release it since the flow is vital for a codelab`

- We've had ~15 users run through it via UX studies.
- I've recruited @maguinis to run through it once the latest UX proposed changes are incorporated.

`Also, a UX person should give it a once over.`
- UX is working extremely closely on this. Thanks!

legalcodes commented:
Thanks for these notes -- will hammer these out asap.

Feel free to leave all your feedback whenever it comes!

legalcodes commented:
I've removed the intro heading, removed the example DartPad, however I didn't add the "text that still needs to remain" back into the codelab anywhere. The only sensible place I could come up with as a possible alternative location for that text would be before a DartPad exercise (since we just removed it from the first DartPad example). However, if that text sits before a DartPad exercise, it exacerbate what is already a tough problem -- the distance a user must scroll between a DartPad exercise and the instructions for that exercise. All thoughts welcome. 

Since the "Key terms" definitions were quick similar in content to the "Quick review" sections, I combined the content wherever necessary and simply removed the "Quick Review" bolded text altogether. 

That leaves only one remaining "Quick terms" section which was not at all duplicative with "Key terms" definitions. All thoughts welcome on this matter too!

kwalrath commented on src/_includes/dartpad-exercise-how-to.md:
```diff
@@ -0,0 +1,16 @@
+The embedded editors in this codelab have partially completed code snippets.
```
Should this really be a separate file? It's currently only used by the cheatsheet codelab.

legalcodes replied:
My thinking is that so long as we are using DartPad, we'll be sharing some instructions / introductory language for it. If I'm mistaken, happy to take this out. I put this here to remind us to consider using common DartPad instructions in the future.

legalcodes replied:
Nevertheless I've removed it

kwalrath commented on src/_includes/dartpad-exercise-how-to.md:
```diff
@@ -0,0 +1,16 @@
+restores the editor to its original state.
+
+<aside class="alert alert-warning">
+The embedded editors use an experimental version of DartPad.
```
I think we can tone this down, but I can do that in a separate PR.

kwalrath commented on src/codelabs/index.md:
```diff
@@ -20,6 +20,10 @@ Use the latest, experimental version of DartPad to learn,
 remind yourself about, or test your knowledge of
 some of the most commonly used, yet unique features of the Dart language.
 
+### [Futures Async & Await](/codelabs/async-await)
```
This title should match the codelab title, either minus "codelab" or with it (modifying the cheatsheet codelab's link text if necessary to be consistent with your decision).

kwalrath commented on src/codelabs/index.md:
```diff
@@ -20,6 +20,10 @@ Use the latest, experimental version of DartPad to learn,
 
+### [Futures Async & Await](/codelabs/async-await)
+
+Use DartPad to learn how to write asynchronous code in Dart via futures and the
```
We should be consistent with how this codelab and the previous one describe DartPad. It *is* a different version than the one in Intro to Dart, but I'm not sure we need to call that out.

via -> using
(https://developers.google.com/style/word-list#letter-V)

legalcodes replied:
It's definitely an interesting thing to think about -- this codelab uses features of DartPad that other codelabs do not, and that may justify different descriptions. Also, we cut down a bunch of the DartPad description language previously because it added too much verbiage up front.

kwalrath commented on src/codelabs/index.md:
```diff
@@ -20,6 +20,10 @@ Use the latest, experimental version of DartPad to learn,
+### [Futures Async & Await](/codelabs/async-await)
+
+Use DartPad to learn how to write asynchronous code in Dart via futures and the
+async/await keywords.
```
/ -> and

legalcodes commented on _config.yml:
```diff
@@ -92,6 +94,30 @@ assets:
+
+os-list: [Windows, macOS, Linux]
+
+alert:
```
Looks like it does!

legalcodes commented on .travis.yml:
```diff
@@ -66,4 +66,4 @@ deploy:
 # Only run Travis jobs for named branches (to avoid double builds for each PR)
 branches:
-  only: [master, /^\d\.x$/, /^\d-dev$/, /travis-build/]
+  only: [master, /^\d\.x$/, /^\d-dev$/, /travis-build/]
```
Not sure how this crept in, should be fixed when I update this PR w/ some changes you've pointed out.

legalcodes commented on src/_data/side-nav.yml:
```diff
@@ -12,6 +12,8 @@
           permalink: /codelabs
         - title: Cheatsheet codelab
           permalink: /codelabs/dart-cheatsheet
+        - title: Asynchronous Programming with Dart
```
I like the second option since it avoids repetition of "Codelab" in sub-menu items that are members of the "Codelab" menu item, will make this change. I'm assuming I should change the title of the cheatsheet as well -- if not, let me know!

legalcodes replied:
On second glance, it doesn't look like we are currently matching codelab titles to their menu item counterparts, so I'll leave the codelab titles alone.

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+tests pass. 
+You don't need to implement `main()`.
+
+To simulate asynchronous operations, your code will call the following functions, which are 
```
your code will call -> call

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+{% endprettify %}
+#### Part 1: `reportUserRole()`
+
+Add logic to the `reportUserRole()` function:
```
"Add logic" doesn't sound right, and it could be easy to misread the following bullets as being already true, rather than goals. Maybe:

Add code to the `reportUserRole` function so that it does the following:

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+
+Add logic to the `reportUserRole()` function:
+<!-- Some bulleted items are intentionally lacking punctuation to avoid confusing the users about characters in string values -->
+* `reportUserRole()` returns a future that completes with the String: `"User role: <user role>"`
```
Returns a future that completes with the following string: <code>"User role: <em>&lt;user role></em>"</code>

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+Add logic to the `reportUserRole()` function:
+<!-- Some bulleted items are intentionally lacking punctuation to avoid confusing the users about characters in string values -->
+* `reportUserRole()` returns a future that completes with the String: `"User role: <user role>"`
+  * Example return value from `reportUserRole()`: `"User role: tester"` 
```
delete "from `reportUserRole()`"

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+<!-- Some bulleted items are intentionally lacking punctuation to avoid confusing the users about characters in string values -->
+* `reportUserRole()` returns a future that completes with the String: `"User role: <user role>"`
+  * Example return value from `reportUserRole()`: `"User role: tester"` 
+  * Note: you must use the actual value returned by `getRole()`; copying and pasting the example return value won't make the test pass. 
```
you -> You

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+* `reportUserRole()` returns a future that completes with the String: `"User role: <user role>"`
+  * Example return value from `reportUserRole()`: `"User role: tester"` 
+  * Note: you must use the actual value returned by `getRole()`; copying and pasting the example return value won't make the test pass. 
+* Get the user role by calling the provided function `getRole()` 
```
Get -> Gets
Add a period at the end of this sentence.

kwalrath replied:
Reverse the two bullets?

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+* Get the user role by calling the provided function `getRole()` 
+
+####  Part 2: `reportLogins()`
+Implement an `async` function `reportLogins()`:
```
Do similar changes to this part.

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+</iframe>
+
+  {{ site.alert.info }}
+    If your tests pass, you can ignore [Info-level messages.](/guides/language/analysis-options#customizing-analysis-rules)
```
Info -> info

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+  {{ site.alert.end }}
+
+## Handling errors
+Handle errors in an `async` function by using try-catch: 
```
Handle -> To handle
function by using -> function, use

kwalrath commented on src/codelabs/async-await/index.md:
````diff
@@ -0,0 +1,473 @@
+  }
+```
+Within an `async` function, you can write [try-catch clauses](/guides/language/language-tour#catch)
+the same way for asynchronous and synchronous code. 
````
for asynchronous and -> you would in 
?

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+  src="https://dartpad.dartlang.org/experimental/embed-new-dart.html?id=25ade03f0632878a9169209e3cd7bef2"
+  style="border: 1px solid lightgrey;"
+  frameborder="no"
+  height="475"
```
Make this a scooch taller?

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+asynchronous function. Can you guess what the output will be before running it?
+<!-- [//]: https://gist.github.com/25ade03f0632878a9169209e3cd7bef2 -->
+<iframe 
+  src="https://dartpad.dartlang.org/experimental/embed-new-dart.html?id=25ade03f0632878a9169209e3cd7bef2"
```
The code says "Imagine that this function... throws an exception." But it does throw an exception!!! Change to "... more complex." ?

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+
+### Exercise: Practice handling errors
+
+<!-- TODO: same fix as above w/ section about provided functions -->
```
Same general comments as above. Also, add an intro here.

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+### Exercise: Putting it all together
+
+It's time to practice what you've learned in one final exercise. 
+To simulate asynchronous operations, __this exercise provides the async functions
```
clean up this description, and apply the comments from above.

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+</iframe>
+
+{{ site.alert.info }}
+  Dart can [infer the return type](https://dart.dev/guides/language/sound-dart#type-inference) for you.
```
Need a little context here.

legalcodes replied:
Some users didn't know about type inference but wound up encountering it one way or another, and in many cases it was helpful to have a link somewhere when it arose. There's no specific context with respect to the content of the codelab, it's just something that came up. It can come up any number of ways, there's no example that implicated it more than others.

kwalrath replied:
Even so, you need a brief reference to some code, or else "infer the return type" means nothing.

legalcodes replied:
Do you mean something like: 

`Instead of adding Future<String> to your function, Dart can infer the return type for you...`

kwalrath replied:
Yes, something like that. But only if `Future<String>` is already visible.

Also, the actor should be clearer. Here's one possibility: 

You might have noticed that the functions in the exercises don't have return types. That's because Dart can infer the return type for you. Omitting return types is fine when you're prototyping, but when you write production code, we recommend that you specify the return type.

legalcodes replied:
Okay, makes sense

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+
+## What's next?
+
+* [Asynchronous programming: futures & async-await](/tutorials/language/futures) tutorial
```
These bullet items should be parallel

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+
+{% prettify dart %}
+getRole() → Future<String> 
+// Returns a short description of the user's role
```
Sure, will make this change. I'll simply note that I tried this out early on in the life of the codelab, and UX gave some feedback that part of the exercise is to ingrain using the `Future<String>` part, which we're giving that away here. Happy to use a table.

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+---
+title: "Asynchronous Dart codelab: futures, async, and await"
+description: A series of explanations and exercises for writing asynchronous code in Dart
```
When I wrote this, I was taking a look at the cheatsheet description: 
`Interactively learn (or relearn) some of Dart's unique features.`

Which also isn't a complete sentence. Shouldn't our descriptions be parallel (i.e. they are either complete sentences or not?) If so, should I change the cheatsheet description as well?

kwalrath replied:
It would be nice if they were parallel, but it isn't essential, since they're usually in isolation. 

But actually, that is a complete sentence. :) It's just elliptical.

legalcodes replied:
That's great! If we don't need them to be parallel, I've gone with: 
`Learn about and practice writing asynchronous code in DartPad!` 
But feel free to throw down some other compelling descriptions if you have something in mind.

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+title: "Asynchronous Dart codelab: futures, async, and await"
+description: A series of explanations and exercises for writing asynchronous code in Dart
+date: 2019-06-12
+tags: [ future, async, await, asynchronous, try, catch, error ]
```
Sure, removed.

legalcodes replied:
It's really up to us how we use metadata for our static content. IMHO the out-of-the-box metadata jekyll exposes via [frontmatter](https://jekyllrb.com/docs/front-matter/) (categories, tags, date) is useful -- if not now, probably in the future. It's much easier to add as we publish each new article/page/codelab than all at once later if we decide we want it.

kwalrath replied:
I'm just worried about maintaining a bunch of stuff that doesn't get used. Or worse, that _is_ used, but not consistently, so that the results in whatever uses it are weird. E.g. would it be weird to have this page tagged "try" and "catch", but not have those tags in the language tour?

legalcodes replied:
I agree that, whatever we decide, we should be intentional about it. Perhaps this warrants creating an issue so we discuss it sooner rather than later.

kwalrath replied:
Please do.

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+---
+title: "Asynchronous Dart codelab: futures, async, and await"
+description: A series of explanations and exercises for writing asynchronous code in Dart
+date: 2019-06-12
```
Deleted.

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+A future of type `Future<T>` completes with a value of type `T`.
+For example, `Future<String>` is the type signature for a
+future that produces a string value. If a future doesn't produce a usable value
+– for example, if you only need to log that a task is complete – then the future's type is `Future<void>`.
```
I agree, I've removed the "example" bit.

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+If the asynchronous operation performed by the function fails for any reason, the 
+future completes with an error.
+
+### Example: Introducing futures
```
I'm happy to riff on other possible names: 
```
Example: 
1. Fun with Futures
2. Futures Fundamentals
3. The basics of Futures
4. Facts about Futures
5. Back to the Futures - this would require re-working the example to use a lot of "Maaaartty" or "MCFLY!!" strings. Why didn't we notice this sooner? I regret everything.

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+    <h4> Example: synchronous functions</h4>
+
+{% prettify dart %}
+// Synchronous
```
Wait, do we have a way to do that right now?

kwalrath replied:
Yes, we do. We do it a lot in Effective Dart & the language tour, but it's all over the site. Check out https://raw.githubusercontent.com/dart-lang/site-www/master/src/samples/index.md.

kwalrath replied:
There's some doc in https://github.com/dart-lang/site-www/tree/master/examples, which is the directory that contains the included code and (where appropriate) tests.

legalcodes replied:
Nice, see https://github.com/dart-lang/site-www/issues/1810

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+asynchronous.
+* **async function**:  An `async` function is a function labeled with the `async` keyword. 
+* **await**: You can use the `await` keyword to get the completed result of an
+asynchronous expression. You can only use the `await` keyword within an `async` function. 
```
How about: `The await keyword only works within an async function`

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+
+## Execution flow with async and await
+
+As of Dart 2, functions marked as `async` run synchronously until the first
```
Sure, done. 

IMHO the styling for these notes is too loud -- it adds a lot of a visual noise to the page, both because of the margins (eats up screen real estate *really* quickly and because of distracting bg color it adds. I try to avoid using our site.alert  at all costs.

kwalrath replied:
If it looks bad, let's not do it. We've just had experience that having "As of Dart 2" wording confuses readers, so I'd like to avoid doing that. But I'd also like to avoid implying that it's always been this way, so that people who knew the first behavior don't get confused.

legalcodes commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+
+As of Dart 2, functions marked as `async` run synchronously until the first
+`await` keyword. This means that within an `async` function body, all
+synchronous code before the first `await` keyword executes immediately. 
```
Sure, done. 

IMHO using code font for `async` but not for async function risks confusing the reader, or at least visually distracting the reader even if they are able to figure out why it's not in code font in some places.

kwalrath replied:
I'm actually fine with `async` function. We just aren't consistent about it, even within this page. How about you make that change everywhere in the site (or file an issue to do so) and note this decision in go/dart-terms?

legalcodes replied:
Cool, see https://github.com/dart-lang/site-www/issues/1810

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -1,23 +1,20 @@
 
 To get the most out of this codelab, you should have the following:
-* Knowledge of basic Dart syntax.
+* [basic Dart syntax](/codelabs/dart-cheatsheet).
```
I like the idea of providing additional materials to help users catch up. I wonder if this doc (https://dart.dev/samples) works better. This "Language samples" doc is closer to the prerequisites used in the UX study. I feel the cheatsheet codelab has too many distractions that are irrelevant to this future codelab.

kwalrath replied:
Sure, /samples works for me.

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -417,11 +432,12 @@ width="100%" >
 
```
In ToC, it's under "Handling errors". I guess it should be move a level up?

legalcodes replied:
@kwalrath I could be wrong, but I think ToC level for exercises and examples may be a point of disagreement here. If it is, I'll leave it to you two.

galeyang replied:
I'm out next Monday. Feel free to ignore this one if other content are ready to go. It's a minor point. We can revisit later.

kwalrath replied:
Agreed that this exercise doesn't belong under Handling errors. Changing `###` to `##` makes sense here.

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -115,10 +113,10 @@ future completes with an error.
+In the following example, `getUserOrder()` returns a future that completes after
+printing to the console. Because it doesn't return a usable value, 
+`getUserOrder()` has the type `Future<void>`. Before you run the example, 
+try to predict which will print first: "Large Latte" or "Fetching user order..." ?
```
"try.....?" -> should be end with a period.

galeyang commented:
(I can't add inline comments for this one)
In the section: "Example: Completing with an error": "How do you think you might handle the error?" This line throws me a question but doesn't tell me the answer. I'm not sure as a reader, what I'm supposed to do with this instruction and I may not know that I'll learn that later.

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -176,23 +177,27 @@ when using `async` and `await`:
 * __The `await` keyword works only in `async` functions.__
 * __To define an async function, add `async` before the function body:__
```
Remove colon(:)?

galeyang replied:
Oh, I noticed that @kwalrath suggested to add this colon. Hi Kathy, it's a little confusing to me to have this colon at the end of a list <ul>. Also, the content below this list seem to include two points.

kwalrath replied:
The colon isn't supposed to be in the list, it's supposed to be in the intro to the sample.

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -176,23 +177,27 @@ when using `async` and `await`:
 * __The `await` keyword works only in `async` functions.__
```
Reverse the order of these two lines? In the following part, you first talk about async and then await.

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -189,13 +184,24 @@ First, add the `async` keyword before the function body.
+If the function doesn't explicitly return a value, then the return type is `Future<void>`:
+
 {% prettify dart %}
     [!Future<void>!] main() async {
```
It's a little confusing to me that you added Future<void> before the main() but didn't include this in the later side-by-side example. 

The content says "If needed, update the function’s type signature". This makes me wonder "so why did you add a function’s type signature here but didn't add a function’s type signature in "Example: asynchronous functions".

legalcodes replied:
I'm removing the `If needed, update the function's type signature" sentence.

legalcodes replied:
I agree with your point about `you added Future before the main() but didn't include this in the later side-by-side example.`, which is why I think `main()` might not be the best choice for this example. @kathyw thoughts? It seems sort of odd (but not wrong) to show examples with `Future<void> main()` everywhere, but it also seems odd to do it in some places but not others.

galeyang replied:
I'm curious about the rationale of changing from `converts createOrderMessage ()` to `concerts main()`.

kwalrath replied:
I'm trying to remember why we changed from `createOrderMessage()` to `main()`. I agree that it might be better to show something that isn't `main()`, so we don't have to dance around the return type declaration being OK to leave off. (Or we can do the dance later.)

I think the text was inconsistent about `Future<void>` vs. `Future<TypeOfWhatYouReturn>`, so we picked `Future<void>`. Also, maybe we wanted to handle the no-return case? But we could do that separately.

galeyang replied:
> Also, maybe we wanted to handle the no-return case? But we could do that separately.

It seems that both examples in the [What is a future?](https://jt-ww-1.firebaseapp.com/codelabs/async-await#what-is-a-future) section has handled the no-return case using Future<void>.

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,499 @@
+description: Learn about and practice writing asynchronous code in DartPad!
+---
+This codelab teaches you how to write asynchronous code using
+futures and the `async` and `await` keywords. This codelab includes
```
Starting two sentences in a row with "This codelab" isn't ideal. My original suggestion was this:

> This codelab teaches you how to write asynchronous....
> Using the embedded DartPad editors, you can test your...

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,473 @@
+
+Here's an example  that converts `main()` from a synchronous to asynchronous function.
+
+First, add the `async` keyword before the function body.
```
I don't see this change. The style guide favors using a colon if the intro immediately precedes the sample (https://developers.google.com/style/code-samples#intros).

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -150,52 +145,66 @@ How do you think you might handle the error?
-Remember, you can only use the `await` keyword within an `async` function. 
-The following examples provide a comparison of synchronous and asynchronous functions.
-If your window is wide enough, you'll see the two examples side-by-side.
+Now that you have an async function, you can use the `await` keyword to wait
```
async function -> `async` function (global)

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -257,25 +266,34 @@ main() [!async!] {
 
-As of Dart 2, functions marked as `async` run synchronously until the first
-`await` keyword. This means that within an `async` function body, all
+Async functions run synchronously until the first
```
Async functions run -> An `async` function runs

kwalrath commented:
> (I can't add inline comments for this one)
> In the section: "Example: Completing with an error": "How do you think you might handle the error?" This line throws me a question but doesn't tell me the answer. I'm not sure as a reader, what I'm supposed to do with this instruction and I may not know that I'll learn that later.

Maybe say just "A bit later you'll learn how to handle the error."

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -306,35 +324,40 @@ code snippets. Your task is to complete the exercise by writing code to make the
 You don't need to implement `main()`.
 
-To simulate asynchronous operations, your code will call the following functions, which are 
+To simulate asynchronous operations, call the following functions which are 
```
which -> , which

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -306,35 +324,40 @@ code snippets. Your task is to complete the exercise by writing code to make the
-  * Example return value from `reportUserRole()`: `"User role: tester"` 
-  * Note: you must use the actual value returned by `getRole()`; copying and pasting the example return value won't make the test pass. 
-* Get the user role by calling the provided function `getRole()` 
+Add code to the `reportUserRole` function so that it does the following:
```
reportUserRole -> reportUserRole()

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -306,35 +324,40 @@ code snippets. Your task is to complete the exercise by writing code to make the
+Add code to the `reportUserRole` function so that it does the following:
+<!-- Some bulleted items are intentionally lacking punctuation to avoid 
+confusing the users about characters in string values -->
+* `reportUserRole()` Returns a future that completes with the following 
```
delete `reportUserRole()`

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -306,35 +324,40 @@ code snippets. Your task is to complete the exercise by writing code to make the
 ####  Part 2: `reportLogins()`
-Implement an `async` function `reportLogins()`:
+Implement an async function `reportLogins()`:
 * `reportLogins()` returns the string `"Total number of logins: <# of logins>"`.
```
similar changes to above

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -277,21 +277,21 @@ The asynchronous example is different in three ways:
 * **async**: You can use the `async` keyword before a function's body to mark it as
 asynchronous.
-* **async function**:  An async function is a function labeled with the `async` 
+* **async function**:  An `async` function is a function labeled with the `async` 
```
**async function** -> **`async` function**

kwalrath commented:
Uh oh, the build failed (on both stable & dev). Something about the code excerpts, but I didn't think you changed anything in the code. Can you figure this out, @legalcodes? (Or I can look at it tomorrow.)

legalcodes commented:
I'll take look, will see what turns up.

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,498 @@
+### Exercise: Putting it all together
+
+It's time to practice what you've learned in one final exercise. 
+To simulate asynchronous operations, this exercise provides the asynchronous 
```
Delete this sentence and the next one, or combine them somehow with the next paragraph. (As is, it's too repetitive.)

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,498 @@
+* `greetUser()` creates a greeting for the user by calling `addHello()`, 
+passing it the username, and returning the result. 
+  * For example, if the username is "Jenny", `greetUser()` should create and
+   return the greeting "Hello Jenny"
```
Jenny" -> Jenny".
OR
the greeting... -> the following: "Hello Jenny"

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,498 @@
+####  Part 3: `sayGoodbye()` 
+<!-- pull out descriptions of  -->
+* Write a function `sayGoodbye()` that does the following:
+* `sayGoodbye()` takes no arguments.
```
Indent this and the following 2 items.
Delete the function name and capitalize the verb:

* ...following:
  * Takes no arguments.
  * Catches any errors.
  * Calls... 
* If... succeeds, `sayGoodbye()` returns the string...

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,498 @@
+* `sayGoodbye()` calls the provided asynchronous function `logoutUser()`.
+<!-- italicize <result> -->
+* If `logoutUser()` succeeds, `sayGoodbye()` returns the string "\<result\> 
+Thanks, see you next time" where \<result\> is the String value returned by calling `logoutUser()`.
```
returned by calling -> produced (eventually) by `logoutUser()`.

(or something else that acknowledges that logoutUser doesn't return a string value)

galeyang commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,498 @@
+* [Asynchronous programming: futures & async-await](/tutorials/language/futures) tutorial
+* [Dart's type system](/guides/language/sound-dart) (Includes more examples of type signatures with futures).
+* [Preferring signatures in function type annotations](/guides/language/effective-dart/design#prefer-signatures-in-function-type-annotations)
+* [Effective Dart](/guides/language/effective-dart)
```
Does this tutorial https://dart.dev/tutorials/language/futures cover more content than this codelab? If it's mostly repeating, maybe it's "other resources' rather than "next step". 

--
In previous UX study, users said they want to learn more about how Dart handles threads, controlling flow/loop, scheduling, and more complex asynchronous tasks other than grabbing values. It seems like Dart uses `streams`, `Isolates`, `event loops`. Hence, we can consider adding these two docs as next step.

- [Asynchronous programming: streams](https://dart.dev/tutorials/language/streams) tutorial: show you how to work with sequence of asynchronous events.
- [Dart asynchronous programming: Isolates and event loops](https://medium.com/dartlang/dart-asynchronous-programming-isolates-and-event-loops-bffc3e296a6a) (link to an external doc on Dart's Medium)

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -387,7 +387,7 @@ the same way you would in synchronous code.
 ### Example: async and await with try-catch
 Run the following example to see how to handle an error from an 
-asynchronous function. Can you guess what the output will be before running it?
+asynchronous function. What do you think the output will be before running it?
```
delete "before running it"

kwalrath commented on src/codelabs/async-await/index.md:
```diff
@@ -0,0 +1,493 @@
+asynchronous expression. The `await` keyword only works within an `async` function.
+{{ site.alert.end }}
+
+## Execution flow with async and await
```
`##` -> `###`
(I noticed in the side nav that the exercise applies to "Working with futures", not "Execution flow". I think this change fixes that problem.)
